File: backend/repositories/userRepository.js
// backend/repositories/userRepository.js

/**
 * Mendapatkan data user berdasarkan username.
 * @param {import('mysql2/promise').Connection} connection
 * @param {string} username
 * @returns {Promise<Object|null>}
 */
export const getUserByUsername = async (connection, username) => {
  const [rows] = await connection.query("SELECT * FROM users WHERE username = ?", [username]);
  return rows.length > 0 ? rows[0] : null;
};

/**
 * Mendapatkan data user berdasarkan ID.
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} id
 * @returns {Promise<Object|null>}
 */
export const getUserById = async (connection, id) => {
  const [rows] = await connection.query("SELECT * FROM users WHERE id = ?", [id]);
  return rows.length > 0 ? rows[0] : null;
};

/**
 * Mendapatkan role dan daftar permissions untuk seorang user berdasarkan role_id-nya.
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} roleId
 * @returns {Promise<{role: string, permissions: string[]}>}
 */
export const getRoleAndPermissions = async (connection, roleId) => {
  const [roleRows] = await connection.query(
    `
      SELECT r.name as role, p.name as permission
      FROM roles r
      LEFT JOIN role_permission rp ON r.id = rp.role_id
      LEFT JOIN permissions p ON rp.permission_id = p.id
      WHERE r.id = ?
    `,
    [roleId]
  );

  const permissions = roleRows.map((row) => row.permission).filter((p) => p);
  const role = roleRows[0]?.role || "user";

  return { role, permissions };
};

/**
 * Memperbarui profil pengguna (nickname dan/atau password).
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} userId
 * @param {Object} updates
 * @param {string} [updates.nickname]
 * @param {string} [updates.hashedNewPassword]
 * @returns {Promise<void>}
 */
export const updateProfile = async (connection, userId, updates) => {
  const updateFields = [];
  const updateValues = [];

  if (updates.nickname !== undefined) {
    updateFields.push("nickname = ?");
    updateValues.push(updates.nickname);
  }

  if (updates.hashedNewPassword) {
    updateFields.push("password_hash = ?");
    updateValues.push(updates.hashedNewPassword);
  }

  if (updateFields.length > 0) {
    const query = `UPDATE users SET ${updateFields.join(", ")} WHERE id = ?`;
    updateValues.push(userId);
    await connection.query(query, updateValues);
  }
};

/**
 * Mendapatkan lokasi gudang yang dapat diakses oleh user.
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} userId
 * @returns {Promise<Array<Object>>}
 */
export const getUserLocations = async (connection, userId) => {
  const query = `
    SELECT l.id, l.code, l.building, l.floor, l.name
    FROM locations l
    JOIN user_locations ul ON l.id = ul.location_id
    WHERE ul.user_id = ?
    ORDER BY l.id ASC
  `;
  const [locations] = await connection.query(query, [userId]);
  return locations;
};

/**
 * Mendapatkan daftar ID user berdasarkan permission tertentu.
 * @param {import('mysql2/promise').Connection} connection
 * @param {string} permissionName
 * @returns {Promise<number[]>}
 */
export const getUserIdsByPermission = async (connection, permissionName) => {
  const [rows] = await connection.query(
    `
      SELECT u.id 
      FROM users u
      JOIN role_permission rp ON u.role_id = rp.role_id
      JOIN permissions p ON rp.permission_id = p.id
      WHERE p.name = ? AND u.is_active = 1
    `,
    [permissionName]
  );
  return rows.map(r => r.id);
};

File: backend/internal/repository/user_repository.go
package repository

import (
	"context"
	"database/sql"

	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/dps-wmhris/backend/internal/model"
	"github.com/jmoiron/sqlx"
)

// UserRepository defines the interface for user data operations
type UserRepository interface {
	GetAll(ctx context.Context) ([]model.User, error)
	FindByID(ctx context.Context, id int) (*model.User, error)
	FindByUsername(ctx context.Context, username string) (*model.User, error)
	GetRoleAndPermissions(ctx context.Context, roleID int) (string, []string, error)
	UpdateProfile(ctx context.Context, db sqlx.ExtContext, userID int, nickname *string, hashedPassword *string) error
	GetUserLocations(ctx context.Context, userID int) ([]dto.UserLocationResponse, error)
	CheckUserLocationPermission(ctx context.Context, userID int, locationID int) (bool, error)
}

type userRepositoryImpl struct {
	db *sqlx.DB
}

// NewUserRepository injects the database dependency
func NewUserRepository(db *sqlx.DB) UserRepository {
	return &userRepositoryImpl{db: db}
}

func (r *userRepositoryImpl) GetAll(ctx context.Context) ([]model.User, error) {
	var users []model.User
	query := `
		SELECT 
			id, username, nickname, is_active, password_hash, role_id, 
			shift_id, created_at, updated_at, exclude_from_attendance 
		FROM users 
		WHERE is_active = 1
		ORDER BY username ASC`
	
	err := r.db.SelectContext(ctx, &users, query)
	return users, err
}

func (r *userRepositoryImpl) FindByID(ctx context.Context, id int) (*model.User, error) {
	var user model.User
	query := `
		SELECT 
			id, username, nickname, is_active, password_hash, role_id, 
			shift_id, created_at, updated_at, exclude_from_attendance 
		FROM users 
		WHERE id = ?`
	
	err := r.db.GetContext(ctx, &user, query, id)
	if err != nil {
		return nil, err
	}
	return &user, nil
}

func (r *userRepositoryImpl) FindByUsername(ctx context.Context, username string) (*model.User, error) {
	var user model.User
	query := `
		SELECT 
			id, username, nickname, is_active, password_hash, role_id, 
			shift_id, created_at, updated_at, exclude_from_attendance 
		FROM users 
		WHERE username = ? LIMIT 1`
		
	err := r.db.GetContext(ctx, &user, query, username)
	if err != nil {
		return nil, err
	}
	return &user, nil
}

func (r *userRepositoryImpl) GetRoleAndPermissions(ctx context.Context, roleID int) (string, []string, error) {
	query := `
		SELECT r.name as role, p.name as permission
		FROM roles r
		LEFT JOIN role_permission rp ON r.id = rp.role_id
		LEFT JOIN permissions p ON rp.permission_id = p.id
		WHERE r.id = ?
	`
	
	type RolePerm struct {
		Role       string  `db:"role"`
		Permission *string `db:"permission"`
	}
	
	var rows []RolePerm
	err := r.db.SelectContext(ctx, &rows, query, roleID)
	if err != nil {
		return "", nil, err
	}
	
	if len(rows) == 0 {
		return "user", []string{}, nil
	}
	
	role := rows[0].Role
	if role == "" {
		role = "user"
	}
	
	var permissions []string
	for _, row := range rows {
		if row.Permission != nil && *row.Permission != "" {
			permissions = append(permissions, *row.Permission)
		}
	}
	
	return role, permissions, nil
}

func (r *userRepositoryImpl) UpdateProfile(ctx context.Context, db sqlx.ExtContext, userID int, nickname *string, hashedPassword *string) error {
	var updateFields []string
	var args []interface{}
	
	if nickname != nil {
		updateFields = append(updateFields, "nickname = ?")
		args = append(args, *nickname)
	}
	
	if hashedPassword != nil {
		updateFields = append(updateFields, "password_hash = ?")
		args = append(args, *hashedPassword)
	}
	
	if len(updateFields) == 0 {
		return nil
	}
	
	query := "UPDATE users SET "
	for i, field := range updateFields {
		query += field
		if i < len(updateFields)-1 {
			query += ", "
		}
	}
	query += " WHERE id = ?"
	args = append(args, userID)
	
	_, err := db.ExecContext(ctx, query, args...)
	return err
}

func (r *userRepositoryImpl) GetUserLocations(ctx context.Context, userID int) ([]dto.UserLocationResponse, error) {
	query := `
		SELECT l.id, l.code, l.building, COALESCE(CAST(l.floor AS CHAR), '') as floor, COALESCE(l.name, '') as name
		FROM locations l
		JOIN user_locations ul ON l.id = ul.location_id
		WHERE ul.user_id = ?
		ORDER BY l.id ASC
	`
	
	var locations []dto.UserLocationResponse
	err := r.db.SelectContext(ctx, &locations, query, userID)
	if err != nil {
		return nil, err
	}
	if locations == nil {
		locations = []dto.UserLocationResponse{}
	}
	
	return locations, nil
}

func (r *userRepositoryImpl) CheckUserLocationPermission(ctx context.Context, userID int, locationID int) (bool, error) {
	query := "SELECT 1 FROM user_locations WHERE user_id = ? AND location_id = ?"
	var exists int
	err := r.db.GetContext(ctx, &exists, query, userID, locationID)
	if err != nil {
		if err == sql.ErrNoRows {
			return false, nil
		}
		return false, err
	}
	return true, nil
}
